File: src/main/java/org/codeberg/zenxarch/zombies/loot_table/function/EnchantmentProviderLootFunction.java
package org.codeberg.zenxarch.zombies.loot_table.function;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.enchantment.provider.EnchantmentProvider;
import net.minecraft.item.ItemStack;
import net.minecraft.loot.context.LootContext;
import net.minecraft.loot.function.LootFunction;
import net.minecraft.loot.function.LootFunctionType;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import org.codeberg.zenxarch.zombies.difficulty.ExtendedDifficulty;

public record EnchantmentProviderLootFunction(RegistryKey<EnchantmentProvider> provider)
    implements LootFunction {

  public static final MapCodec<EnchantmentProviderLootFunction> CODEC =
      RecordCodecBuilder.mapCodec(
          instance ->
              instance
                  .group(
                      RegistryKey.createCodec(RegistryKeys.ENCHANTMENT_PROVIDER)
                          .fieldOf("provider")
                          .forGetter(EnchantmentProviderLootFunction::provider))
                  .apply(instance, EnchantmentProviderLootFunction::new));

  @Override
  public ItemStack apply(ItemStack stack, LootContext context) {
    EnchantmentHelper.applyEnchantmentProvider(
        stack,
        context.getWorld().getRegistryManager(),
        provider,
        new ExtendedDifficulty(context.getWorld(), context.getLuck(), 0),
        context.getRandom());
    return stack;
  }

  @Override
  public LootFunctionType<? extends LootFunction> getType() {
    return ZombieLootFunctionTypes.ENCHANTMENT_PROVIDER;
  }
}

File: src/main/java/org/codeberg/zenxarch/zombies/loot_table/EnchantmentProviderLootFunction.kt
package org.codeberg.zenxarch.zombies.loot_table

import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.enchantment.provider.EnchantmentProvider
import net.minecraft.item.ItemStack
import net.minecraft.loot.context.LootContext
import net.minecraft.loot.function.LootFunction
import net.minecraft.loot.function.LootFunctionType
import net.minecraft.registry.RegistryKey
import net.minecraft.registry.RegistryKeys
import org.codeberg.zenxarch.zombies.difficulty.ExtendedDifficulty

@JvmRecord
data class EnchantmentProviderLootFunction(val provider: RegistryKey<EnchantmentProvider>) :
    LootFunction {
    override fun apply(stack: ItemStack, context: LootContext): ItemStack {
        EnchantmentHelper.applyEnchantmentProvider(
            stack,
            context.world.registryManager,
            provider,
            ExtendedDifficulty(context.world, context.luck.toDouble(), 0),
            context.random
        )
        return stack
    }

    override fun getType(): LootFunctionType<out LootFunction?> = ZombieLootFunctionTypes.ENCHANTMENT_PROVIDER

    companion object {
        @JvmField
        val CODEC: MapCodec<EnchantmentProviderLootFunction?> =
            RecordCodecBuilder.mapCodec { instance: RecordCodecBuilder.Instance<EnchantmentProviderLootFunction?> ->
                instance.group(
                    RegistryKey.createCodec(RegistryKeys.ENCHANTMENT_PROVIDER).fieldOf("provider")
                        .forGetter { it?.provider }).apply(
                    instance, ::EnchantmentProviderLootFunction
                )
            }
    }
}
